### ISSUE TITLE
Optimization idea for the VM: tail calls

### ISSUE THREAD

**Issue by haberman** [munificent/wren#1058]
I just came across Wren today and it looks very cool. I have been a Lua fan for a while, but I always wished for statically defined classes.

I notice that Wren has a small implementation and also has a focus on speed. For this reason I wanted to point you towards an idea that I think would likely deliver significant performance benefits, and would be reasonably easy to experiment with: https://blog.reverberate.org/2021/04/21/musttail-efficient-interpreters.html

With this approach I think it is possible to implement an interpreter that can match the tight assembly sequences of LuaJIT 2's interpreter without actually having to write in assembly (an example of this is given in the article). This approach is also taken by the wasm3 WebAssembly interpreter.

You'll need `__attribute__((musttail))` (just released in Clang 13) if you want to avoid blowing the stack in debug mode. But with some macros, I think you could potentially structure the code such that you can switch between this and computed goto easily.

Just an idea, feel free to ignore, but I wanted to send it your way. :)

**Comment by mhermier** [comment]
I was thinking about it for a while, but as of current implementation it has quite some implications.
1. In the VM it would means we loose ~32 opcodes values (probably more to adjust stack attributes?) and need to branch call code or duplicate it.
2. In the compiler, it would means we need to add support for possibly "tail" and "musttail" attributes (or what ever solution we adopt), where it would be safe to ignore them.
3. It can cut of stack layers, making stack trace harder to follow when enabled, leading to a fragmentation in usage.

Because of the first point, I don't think it is a good idea to accept such change without good numbers in benchmarks.
In addition, wren does not really have a stack size limit (as of now), so one should not really hit a stack overflow unless he messed up and triggered an infinite recursion. But enabling it would allow to reduce the stress on stack memory allocation.

It is a nice feature that would definitively have a great potential to speed a the VM a lot, but 3rd point still tickles me too much unless we can prove some solid improvements in speeds like more than 2 times faster.

**Comment by mhermier** [comment]
This would change the structure of the dispatch function by creating a
plethora of functions. While there would be functional change, what I
understand of the current situation is that the code fits nicely in a few
cache pages making the code fast, to the point that even tweaking some
opcodes can trigger some dramatic performance changes (at least on my
system). So I suspect this change would have unpredictable performance
implications.

In addition it would rely on the presence of a non standard extension, that
would not resolve nicely if ignored since it would stack overflow the host.

While I understand the motivation for larger projects, I think relying on
such kind of extension is not the smartest move since it reduce the
compiler list. And there are cases, in particular if your project is
embedded, where one has to use an exotic compiler with limited language
support, and usually poor extension support.

**Comment by mhermier** [comment]
It is not a question of appeal. Unless you prove my arguments to be wrong,
the trade of seems not worth as this. If the impact in terms of code
organization is low/null and performance is good enough (let's say ~10%
more bad in worse case scenarios), we can start considering supporting it.

As this, fully supporting it would rely on 2 compiler extensions (inner
functions and tailforce) and this it is my biggest point against the
adoption. But if performance is better than my expectations, it is really
worth it. In particular if it also brings a little bit of stability when
tweaking opcodes.

**Comment by joshgoebel** [comment]
Yes, it seems to me like if the patch was simple and non-invasive (if it would somehow redefine existing macros, as mentioned above) - and made Wren measurably faster on some platforms and more stable (in terms of performance from patch to patch) then if it was an optional compiler flag that'd be pretty sweet.

Looking forward to seeing a PR with some benchmarks if you find the time, @haberman
